pyodata/v2/model.py

```diff
@@ -840,135 +955,165 @@ def from_etree(schema_nodes):
 
-                    # Check if dependent role properties exist
-                    role_name = dependent_role.name
-                    entity_type_name = assoc.end_by_role(role_name).entity_type_name
-                    schema.check_role_property_names(dependent_role, entity_type_name, namespace)
+                            etype = schema.entity_type(end_role.entity_type_info.name, end_role.entity_type_info.namespace)
 
-                decl.associations[assoc.name] = assoc
+                            end_role.entity_type = etype
+                        except KeyError:
+                            raise PyODataModelError(
+                                f'EntityType {end_role.entity_type_info.name} does not exist in Schema '
+                                f'Namespace {end_role.entity_type_info.namespace}')
+
+                    if assoc.referential_constraint is not None:
+                        role_names = [end_role.role for end_role in assoc.end_roles]
+                        principal_role = assoc.referential_constraint.principal
+
+                        # Check if the role was defined in the current association
+                        if principal_role.name not in role_names:
+                            raise RuntimeError(
+                                'Role {} was not defined in association {}'.format(principal_role.name, assoc.name))
+
+                        # Check if principal role properties exist
+                        role_name = principal_role.name
+                        entity_type_name = assoc.end_by_role(role_name).entity_type_name
+                        schema.check_role_property_names(principal_role, entity_type_name, namespace)
+
+                        dependent_role = assoc.referential_constraint.dependent
+
+                        # Check if the role was defined in the current association
+                        if dependent_role.name not in role_names:
+                            raise RuntimeError(
+                                'Role {} was not defined in association {}'.format(dependent_role.name, assoc.name))
+
+                        # Check if dependent role properties exist
+                        role_name = dependent_role.name
+                        entity_type_name = assoc.end_by_role(role_name).entity_type_name
+                        schema.check_role_property_names(dependent_role, entity_type_name, namespace)
+                except (PyODataModelError, RuntimeError) as ex:
+                    config.err_policy(ParserError.ASSOCIATION).resolve(ex)
+                    decl.associations[assoc.name] = NullAssociation(assoc.name)
+                else:
+                    decl.associations[assoc.name] = assoc
 
         # resolve navigation properties
         for stype in schema.entity_types:
             # skip collections
-            if stype.is_collection:
+            if isinstance(stype, NullType) or stype.is_collection:
```

Comment:
Please, do not join unrelated expressions with logical operators.

Reply:
danone